src/rest/resource.cpp

```diff
@@ -518,31 +513,22 @@ void Resource::UpdateDiag(std::string aKey, std::vector<otNetworkDiagTlv> &aDiag
 
 void Resource::Diagnostic(const Request &aRequest, Response &aResponse) const
 {
-    otbrError error = OTBR_ERROR_NONE;
+    otThreadSetReceiveDiagnosticGetCallback(mInstance, &Resource::DiagnosticResponseHandler,
+                                            const_cast<Resource *>(this));
     OT_UNUSED_VARIABLE(aRequest);
     struct otIp6Address rloc16address = *otThreadGetRloc(mInstance);
     struct otIp6Address multicastAddress;
 
     VerifyOrExit(otThreadSendDiagnosticGet(mInstance, &rloc16address, kAllTlvTypes, sizeof(kAllTlvTypes)) ==
-                     OT_ERROR_NONE,
-                 error = OTBR_ERROR_REST);
-    VerifyOrExit(otIp6AddressFromString(kMulticastAddrAllRouters, &multicastAddress) == OT_ERROR_NONE,
-                 error = OTBR_ERROR_REST);
+                 OT_ERROR_NONE);
+    VerifyOrExit(otIp6AddressFromString(kMulticastAddrAllRouters, &multicastAddress) == OT_ERROR_NONE);
     VerifyOrExit(otThreadSendDiagnosticGet(mInstance, &multicastAddress, kAllTlvTypes, sizeof(kAllTlvTypes)) ==
-                     OT_ERROR_NONE,
-                 error = OTBR_ERROR_REST);
+                 OT_ERROR_NONE);
 
 exit:
 
-    if (error == OTBR_ERROR_NONE)
-    {
-        aResponse.SetStartTime(steady_clock::now());
-        aResponse.SetCallback();
-    }
-    else
-    {
-        ErrorHandler(aResponse, HttpStatusCode::kStatusInternalServerError);
-    }
+    aResponse.SetStartTime(steady_clock::now());
+    aResponse.SetCallback();
```

Comment:
Don't need to address in this PR: the function name `SetCallback` is bad... I would expect this function accepts a callback function as an argument but it is actually a flag indicates whether we have/enable callback. Can we rename to `SetHasCallback`?

Reply:
> > This PR mainly fixes some issues on current tests of OTBR-REST, and makes some preparations for #532 .
> > 
> > * set timeout for rest-test in case it takes too much time
> > * some modifications on rest-test principles & HTTP response, so this test will work properly when more error code of underlying APIs are defined
> > * bind `ReceiveDiagnosticGetCallback` to `resource` each time when `diagnostics API` is called(consistent with rest-test modifications and more logistic)
> 
> Thanks for the fix. I merged this PR to #532 and it did pass all the checks.
> But I am not clear what exactly does this PR fix. Any detailed explanation? @tttttangTH

 After #532 is applied, several issues occurs.

- some Openthread API call fail for `detached`.  So I add basic configuration for the node before rest-test start.
- one  Openthread API still fails for `no buffer`. So I ignore this error currently (It does fail for an Openthread API call but I don't think it is a failure for our `diagnostics` API ), and will complete this part later after #532, 

- rest-test takes too much time, gives too much log.  It's strange, seems it is result from your PR, when I start the node, then stop `otbr-agent` and restart it again, many logs keep appearing and never stop. This is why we have too much logs in rest-test.  After `factoryreset`, it is back to normal. so I also add a `factoryreset` operation to the node before rest-test which leads to our discussion on callback bind.

Reply:
Another point, I think `HasCallback` can be effectively represent by the nullness of a `mCallback` (a function pointer) member. If `mCallback` is `nullptr`, then we have a Callback, otherwise we don't have. Why do we need something like `HasCallback` ?